scripts/publish_crates-io.py

```diff
@@ -0,0 +1,138 @@
+    else:
+        return f"{crate[0:2]}/{crate[2:4]}/{crate}"
+
+
+def crates_io_published_versions(crate: str) -> List[str]:
+    """
+    Fetches the current list of all published versions of a crate on crates-io.
+
+    Notably, this _includes_ the yanked versions
+    """
+    index_url = f"https://index.crates.io/{index_url_path(crate)}"
+    print(f"Requesting {index_url}")
+    request = urllib.request.Request(index_url)
+    with urllib.request.urlopen(request) as response:
+        # We ignore anything that comes after the first newline
+        data = json.loads(response.read().decode("utf-8").split("\n", 1)[0])
+        # The response can be either a single json object or an array of json object
+        if isinstance(data, dict):
+            return [data["vers"]]
+        else:
+            return [published["vers"] for published in data]
+
+
+def publish(crate: str, version: str, registry: str):
+    """
+    Publish a crate on crates-io.
+
+    You must be already logged in in order to have this function working.
+    """
+    print(f"Publishing {crate} crate on {registry} in version {version}...", end=" ")
+    try:
+        subprocess.check_output(
+            f"cargo publish --registry {registry} -p {crate}",
+            stderr=subprocess.STDOUT,
+            shell=True,
+        )
+    except subprocess.CalledProcessError as e:
+        print(f"Error during publication:\n{e.output}")
+        sys.exit(1)
+    print("OK!")
+
+
+def is_version_published_on_crates_io(crate: str, version: str) -> bool:
+    """
+    Return true if VERSION of CRATE is already on crates-io.
+    """
+    published_versions = crates_io_published_versions(crate)
+    print(f"Published versions are {published_versions}")
+    return version in published_versions
+
+
+def current_version(crate: str) -> str:
+    """
+    Return the current version of CRATE, according to Cargo reading
+    its own metadata.
+    """
+    try:
+        cargo_metadata = json.loads(
+            subprocess.check_output(
+                f"cargo metadata --format-version 1 --no-deps",
+                shell=True,
+            )
+        )
+        current_crate_meta = next(
+            meta for meta in cargo_metadata["packages"] if meta["name"] == crate
+        )
+        return current_crate_meta["version"]
+    except subprocess.CalledProcessError as e:
+        print(f"Cargo error while fetching the current version:\n{e.output}")
+        sys.exit(1)
+
+
+def handle_single_crate_with_crates_io(crate: str):
+    crate_version = current_version(crate)
+    if not is_version_published_on_crates_io(crate, crate_version):
+        publish(crate, crate_version, CRATES_IO)
+
+
+def main(crate: str, registry: str):
+    if registry != CRATES_IO:
+        print("Only crates-io registry is supported now")
+        sys.exit(1)
+
+    if crate == ALL_CRATES:
+        for crate in ALL_CRATES_IN_ORDER:
+            handle_single_crate_with_crates_io(crate)
```

Comment:
We want to introduce a verification that the crate is published before continuing to the next crate. I've noticed that sometimes a takes a little bit for a published crate to be actually available, and continuing early may then result in issues when the next crate depends on the previous.

Reply:
cargo publish will default to polling the registry until the version is available before going to the next one. Maybe that's just happening when a TTY is connected to STDOUT, but that should at least be a flag in publish I could set, I'll check

Reply:
Actually it always waits at most [1 minute for the publication to get propagated](https://github.com/rust-lang/cargo/blob/a59ac81db46fb410de243a2e600e9bbd535b0662/src/cargo/ops/registry.rs#L233-L245), but it's hard to track for how long the change has been there.

At least it's present on the [`rust-1.67.0` branch](https://github.com/rust-lang/cargo/blob/rust-1.67.0/src/cargo/ops/registry.rs#L193) which is current stable, so I'm assuming that it's going to be available in the "stable" toolchain pulled by CI